src/item.cpp

```diff
@@ -4147,6 +4147,91 @@ const item * item::magazine_current() const
     return const_cast<item *>(this)->magazine_current();
 }
 
+bool item::gunmod_compatible( const item& mod, bool alert ) const
+{
+    if( !mod.is_gunmod() ) {
+        debugmsg( "Tried checking compatibility of non-gunmod" );
+        return false;
+    }
+
+    std::string msg;
+
+    if( !is_gun() ) {
+        msg = string_format( _( "That %s is not a weapon." ), tname().c_str() );
+
+    } else if( is_gunmod() ) {
+        msg = string_format( _( "That %s is a gunmod, it can not be modded." ), tname().c_str() );
+
+    } else if( has_gunmod( mod.typeId() ) != -1 ) {
+        msg = string_format( _( "Your %1$s already has a %2$s." ), tname().c_str(), mod.tname( 1 ).c_str() );
+
+    } else if( !type->gun->valid_mod_locations.count( mod.type->gunmod->location ) ) {
+        msg = string_format( _( "Your %s doesn't have a slot for this mod." ), tname().c_str() );
+
+    } else if( get_free_mod_locations( mod.type->gunmod->location ) <= 0 ) {
+        msg = string_format( _( "Your %1$s doesn't have enough room for another %2$s mod." ), tname().c_str(), _( mod.type->gunmod->location.c_str() ) );
+
+    } else if( ammo_remaining() > 0 || magazine_current() ) {
```

Comment:
This does not allow for UPS powered weapons such as the A7 laser rifle, which cannot be unloaded. After firing them once mods cannot be added or removed.